### PR TITLE
 feat: allow extensions to implement CREATE/DROP DATABASE

### PR DESCRIPTION
**Pull request by rjatwal**
Supersedes PR #6075

### CODE REVIEW

File: src/planner/binder/statement/bind_drop.cpp [hunk 1]
```diff
@@ -43,10 +46,33 @@ BoundStatement Binder::Bind(DropStatement &stmt) {
...
+		string database_name = base.name;
+
+		auto &config = DBConfig::GetConfig(context);
+
+		if (config.storage_extensions.empty()) {
+			// attaching and detaching is read-only
+			stmt.info->catalog = SYSTEM_CATALOG;
+		}
+		// for now assume only one storage extension provides the custom drop_database impl
+		for (auto &extension_entry : config.storage_extensions) {
```
Comment (Mytherin) [inline src/planner/binder/statement/bind_drop.cpp:61]: This is already used by DuckDB for `DETACH` - the storage extension does not seem to have any effect here. It creates a plan (sets `result.plan`) which then gets overriden by the regular `DETACH` afterwards.

Having the same code path handle two different cases (`detach` and `drop database`) also seems very error-prone, and might lead to problems where calling `detach` can lead to a database being dropped instead.

**Review by Mytherin — COMMENTED** [review]
Thanks for the changes! One more thing that I did not realize before - the code for `DROP DATABASE` seems to work with the existing `DETACH` code. These are two distinct concepts that should be kept separate in my opinion. Perhaps we can move the `DETACH` code to a separate operator instead?

The PR also seems to be missing code that extends the grammar with support for `DROP DATABASE`. Should that not be added as well?

File: src/planner/binder/statement/bind_drop.cpp (hunk 1 above)
Comment (bleskes) [reply src/planner/binder/statement/bind_drop.cpp:61]: Just to make sure I understand - IMO `detach` is something that's fully DuckDB owned (as it changes the databased registrations) and extension can be notified about it. `drop` (which implies detach) is something that storage extensions are *supposed* to implement as it requires custom deletion. @Mytherin @rjatwal am I thinking about this correctly?

Comment (rjatwal) [reply src/planner/binder/statement/bind_drop.cpp:61]: Yes. Mark and I discussed this being part of the changes in this PR

File: src/parser/transform/statement/transform_detach.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#include "duckdb/common/string_util.hpp"
+
+namespace duckdb {
+
+unique_ptr<DetachStatement> Transformer::TransformDetach(duckdb_libpgquery::PGNode *node) {
+	auto stmt = reinterpret_cast<duckdb_libpgquery::PGDetachStmt *>(node);
+	auto result = make_unique<DetachStatement>();
+	auto info = make_unique<DetachInfo>();
+	info->name = stmt->db_name;
+	info->if_exists = stmt->missing_ok;
```
Comment (rjatwal) [inline src/parser/transform/statement/transform_detach.cpp:13]: Planning to change `missing_ok` => `if_exists`

File: src/common/enums/logical_operator_type.cpp [hunk 3]
```diff
@@ -100,6 +100,8 @@ string LogicalOperatorToString(LogicalOperatorType type) {
 		return "CREATE_SCHEMA";
 	case LogicalOperatorType::LOGICAL_ATTACH:
 		return "ATTACH";
+	case LogicalOperatorType::LOGICAL_DETACH:
+		return "ATTACH";
```
Comment (stephaniewang526) [inline src/common/enums/logical_operator_type.cpp:104]: Is this supposed to return "DETACH"?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**